velox/exec/tests/PartitionAndSerialize.cpp

```diff
@@ -0,0 +1,163 @@
+      stream << field->name();
+    } else if (
+        auto constant =
+            std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr)) {
+      stream << constant->toString();
+    } else {
+      stream << expr->toString();
+    }
+  }
+  stream << ") " << numPartitions_;
+}
+
+namespace {
+class PartitionAndSerializeOperator : public Operator {
+ public:
+  PartitionAndSerializeOperator(
+      int32_t operatorId,
+      DriverCtx* FOLLY_NONNULL ctx,
+      const std::shared_ptr<const PartitionAndSerializeNode>& planNode)
+      : Operator(
+            ctx,
+            planNode->outputType(),
+            operatorId,
+            planNode->id(),
+            "PartitionAndSerialize") {
+    auto inputType = planNode->sources()[0]->outputType();
+    auto keyChannels = toChannels(inputType, planNode->keys());
+
+    partitionFunction_ = std::make_unique<HashPartitionFunction>(
+        planNode->numPartitions(), inputType, keyChannels);
+  }
+
+  bool needsInput() const override {
+    return !input_;
+  }
+
+  void addInput(RowVectorPtr input) override {
+    input_ = std::move(input);
+  }
+
+  RowVectorPtr getOutput() override {
+    if (!input_) {
+      return nullptr;
+    }
+
+    auto numInput = input_->size();
+
+    // TODO Reuse output vector.
+    auto output = std::dynamic_pointer_cast<RowVector>(
+        BaseVector::create(outputType_, numInput, pool()));
+
+    computePartitions(*output->childAt(0)->asFlatVector<int32_t>());
+
+    serializeRows(*output->childAt(1)->asFlatVector<StringView>());
+
+    input_.reset();
+
+    return output;
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return noMoreInput_;
+  }
+
+ private:
+  void computePartitions(FlatVector<int32_t>& partitionsVector) {
+    auto numInput = input_->size();
+
+    partitions_.resize(numInput);
+    partitionFunction_->partition(*input_, partitions_);
+
+    // TODO Avoid copy.
+    partitionsVector.resize(numInput);
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    std::memcpy(rawPartitions, partitions_.data(), sizeof(int32_t) * numInput);
+  }
+
+  void serializeRows(FlatVector<StringView>& dataVector) {
+    auto numInput = input_->size();
+
+    dataVector.resize(numInput);
+
+    // Compute row sizes.
+    rowSizes_.resize(numInput);
+
+    size_t totalSize = 0;
+    for (auto i = 0; i < numInput; ++i) {
+      size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSizeRow(
+          input_->type(), input_.get(), i);
+      rowSizes_[i] = rowSize;
+      totalSize += rowSize;
+    }
+
+    // Allocate memory.
+    auto buffer = dataVector.getBufferWithSpace(totalSize);
+
+    // Serialize rows.
+    auto rawBuffer = buffer->asMutable<char>();
+    size_t offset = 0;
+    for (auto i = 0; i < numInput; ++i) {
+      dataVector.setNoCopy(i, StringView(rawBuffer + offset, rowSizes_[i]));
```

Comment:
Why do we need ```setNoCopy``` here?

Reply:
FlatVector<StringView>::set() method copies the StringView into internal string buffers. We don't want this code, hence, using setNoCopy.